File: 897.递增顺序搜索树.py
#
# @lc app=leetcode.cn id=897 lang=python3
#
# [897] 递增顺序搜索树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def increasingBST(self, root: TreeNode) -> TreeNode:
        all_val = []

        def inOrder(root):
            if root:
                inOrder(root.left)
                all_val.append(root.val)
                inOrder(root.right)

        inOrder(root)

        pre_head = TreeNode()
        curr_node = pre_head

        for val in all_val:
            curr_node.right = TreeNode(val=val)
            curr_node = curr_node.right

        return pre_head.right

# @lc code=end

File: 897.递增顺序搜索树.go
/*
 * @lc app=leetcode.cn id=897 lang=golang
 *
 * [897] 递增顺序搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func increasingBST(root *TreeNode) *TreeNode {
	allNodeVal := []int{}
	var inOrder func(root *TreeNode)
	inOrder = func(root *TreeNode) {
		// 中序遍历将所有 Val 写入 allNodeVal，由于是二叉搜索树，所以不担心有序问题
		if root != nil {
			inOrder((*root).Left)
			allNodeVal = append(allNodeVal, root.Val)
			inOrder((*root).Right)
		}
	}

	inOrder(root)

	preHead := &TreeNode{} // 辅助节点
	currNode := preHead    // 工作节点

	for _, val := range allNodeVal {
		currNode.Right = &TreeNode{Val: val} // 只有右子树
		currNode = currNode.Right
	}

	return preHead.Right
}

// @lc code=end

